src/core/src/runtime/itensor.cpp

```diff
@@ -0,0 +1,279 @@
+                         OPENVINO_ASSERT(allocator, "Allocator was not initialized");
+                         return const_cast<Allocator&>(allocator).allocate(element_type.size() * shape_size(shape));
+                     }()},
+          m_allocator{allocator} {}
+
+    ~AllocatedTensor() {
+        m_allocator.deallocate(m_ptr, get_byte_size());
+    }
+
+    void set_shape(ov::Shape new_shape) override {
+        auto old_byte_size = get_byte_size();
+        m_shape = std::move(new_shape);
+        if (get_byte_size() > old_byte_size) {
+            m_allocator.deallocate(m_ptr, old_byte_size);
+            m_ptr = m_allocator.allocate(get_byte_size());
+        }
+        update_strides();
+    }
+
+private:
+    Allocator m_allocator;
+};
+
+/**
+ * @brief Creates allocated tensor
+ *
+ * @param element_type Tensor element type
+ * @param shape Tensor shape
+ * @param allocator Tensor allocator
+ *
+ * @return Shared pointer to tensor interface
+ */
+std::shared_ptr<ITensor> make_tensor(const element::Type element_type, const Shape& shape, const Allocator& allocator) {
+    return std::make_shared<AllocatedTensor>(element_type, shape, allocator);
+}
+
+/**
+ * @brief ROI tensor on other tensor
+ * ROI tensor holds the owner
+ */
+class RoiTensor : public ITensor {
+public:
+    RoiTensor(const std::shared_ptr<ITensor>& owner, const Coordinate& begin, const Coordinate& end)
+        : m_owner{owner},
+          m_offsets{begin} {
+        OPENVINO_ASSERT(owner->get_element_type().bitwidth() >= 8,
+                        "ROI Tensor for types with bitwidths less then 8 bit is not implemented. Tensor type: ",
+                        owner->get_element_type());
+        auto owner_shape = owner->get_shape();
+        OPENVINO_ASSERT(owner_shape.size() == begin.size());
+        OPENVINO_ASSERT(begin.size() == end.size());
+        m_shape.resize(begin.size());
+        for (size_t i = 0; i < begin.size(); ++i) {
+            OPENVINO_ASSERT(begin[i] <= owner_shape[i]);
+            OPENVINO_ASSERT(end[i] <= owner_shape[i]);
+            m_shape[i] = end[i] - begin[i];
+            OPENVINO_ASSERT(m_shape[i] <= owner_shape[i]);
+        }
+    }
+
+    const element::Type& get_element_type() const override {
+        return m_owner->get_element_type();
+    }
+
+    const Strides& get_strides() const override {
+        return m_owner->get_strides();
+    }
+
+    const Shape& get_shape() const override {
+        return m_shape;
+    }
+
+    void set_shape(ov::Shape new_shape) override {
+        OPENVINO_THROW("Shapes cannot be changed for ROI Tensor");
+    }
+
+    void* data(const element::Type& element_type) const override {
+        auto owner_data = m_owner->data(element_type);
+        auto& strides = get_strides();
+        size_t byte_offset =
+            std::inner_product(m_offsets.begin(), m_offsets.end(), strides.begin(), static_cast<size_t>(0));
+        return static_cast<uint8_t*>(owner_data) + byte_offset;
```

Comment:
should we precompute these values and always store a single value `m_offset` instead of `m_offsets`?

Because in case `data()` is called in a loop, we will have better performance.